Complete code refactor

RodrigoAroeira opened the pull request:
Code was separated into their own files.
### Genetic Algorithm Implementation:

* [`constants.py`](diffhunk://#diff-cfa78e49e2499e95798c1595f053646e7ff560961f38d907d4379082a0b05023R1-R2): Added `KONAMI_CODE` and `GENES` to the constants file and variable names are now uppercase to indicate they're constants.

* [`helpers.py`](diffhunk://#diff-0a56e65e9f7b6e23945f47047693792335c732b570a9a25268e910d38ced0e23R1-R42): 
`populate` function: changed random.randint to random.choice and created a list with the genes instead of their indexes.
`select` function: Instead of sorting original list, returned a new sorted list.
`crossover` function: was refactored, so code is more readable and easier to understand

* [`main.py`](diffhunk://#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R1-R59): Created the main script to run the simulation without all the functions cluttering the file.

### Configuration and Player Representation:

* [`parser.py`](diffhunk://#diff-df00b01568933a06c611778d2d70679891ebf6e950241d03bb2aa27f3e196fe0R1-R29): Added argument parsing to allow configuration of the simulation parameters such as population size, fitness cutoff, mutation rate, win percentage, and maximum iterations.

* [`player.py`](diffhunk://#diff-7ba79b66f88895a6256325cf87311132a41f1b5708cbb699d25b3aa82de9d2f6R1-R43): Defined the `Player` class as a dataclass. Player.lives is no longer needed. Can be removed
`Player.mutate` directly alters Player.dna instead of returning a new one.
`Player.test_fitness` now uses enumerate for easier checking. It now checks total genes in common instead of genes in order. Can be easily reverted back, while also not needing self.lives

RodrigoAroeira committed:
Complete code refactor

maxcan7 reviewed (changes requested):
This looks really good!

My only request would be that for such a substantive change, I'd really like to have some unit tests.

I should have created unit tests in the first place. I should have some time, I may be able to develop some of the unit tests myself or scaffold them, not intending to put it all on you, but I do want that first before we merge this.

I appreciate the work you've put into this, thanks for the PR :)!

maxcan7 commented:
Ok I just pushed a couple commits to main, including a bug fix and a unit tests file (which was how I caught the bug ;)).

Given the nature of this refactor, even if you merge the changes from main into your fork, you'll still need to manually make the change, but I kept the commit atomic so it should be easy to identify.

Likewise, it shouldn't be too difficult to adapt the unit tests for your refactor.

RodrigoAroeira committed:
Merge branch 'main' of https://github.com/maxcan7/genetic_algorithm_konami_code

RodrigoAroeira committed:
refactor: Player.test_fitness now acts more like the original

RodrigoAroeira committed:
chore: update .gitignore with generalized exclusions

RodrigoAroeira committed:
refactor: Changed test to use new refactored code

RodrigoAroeira committed:
refactor: removed unused class attribute Player.lives

RodrigoAroeira committed:
chore: Added requirements.txt file

RodrigoAroeira commented:
I managed to pull your changes to my fork, and I managed to get your unit tests aswell! I now changed the mutate function and made the test_fitness act like the one you originally created, but without needing Player.lives. 

I also changed the unit test. I changed it to use the new refactored code, and removed any Player.lives assertion

I also improved the .gitignore file and added a requirements.txt file.

If I think anything more can be changed, I'll let you know!

maxcan7 commented on test_genetic_algorithm_konami_code.py:
```diff
@@ -1,35 +1,32 @@
 import random
```
Ok we just need to break this test script to correspond to how you've broken up the original main script, and then also should probably add some tests for the parser, but after that we can probably merge, and then I might push a few more commits after that.

RodrigoAroeira replied:
I'm going to push to my fork an updated test file with some tests for the parser stuff. 

Now, for the existing tests you created, I already changed the imports to use the new files, and the tests are still passing.

maxcan7 replied:
ya I just mean, I usually like to have the test scripts broken up so that they correspond to the files of the functions. When it was all just `genetic_algorithm_konami_code.py` it made sense to put all the tests in `test_genetic_algorithm_konami_code.py`, but now that it's broken up across main, player, parser, and helpers, the tests should reflect that in `test_player`, `test_helpers`, etc.

After that it's good to go :).

RodrigoAroeira committed:
test: Added tests for the parser

RodrigoAroeira committed:
chore: Code broken up into their own files.
